Add support for dynamic network policy compilation

jackkleeman opened the pull request:
We compile network policies with the same label into single network
policies with the same effect, to reduce the number of calico policies.
All labelled policies must have the same namespace and selector.

jackkleeman committed:
Add support for dynamic network policy compilation

We compile network policies with the same label into single network
policies with the same effect, to reduce the number of calico policies.
All labelled policies must have the same namespace and selector.

robbiev commented on pkg/controllers/networkpolicy/policy_controller.go:
```diff
@@ -120,7 +153,40 @@ func NewPolicyController(ctx context.Context, clientset *kubernetes.Clientset, c
+
+			if dynamic, ok := obj.(*networkingv1.NetworkPolicy).Labels[dynamicLabel]; ok {
+				// was part of a dynamic policy before deletion
+				toCombine, _ := indexer.ByIndex("combiner", dynamic)
```
Am I right that this will include the one that's currently being deleted? Would this mean we then update a policy to be empty instead of deleting it? (When it's the last one)

jackkleeman replied:
It doesn't include the one being deleted. These functions are always run after the index is updated.

robbiev replied:
Cool, this could be worth commenting.

robbiev commented on pkg/controllers/networkpolicy/policy_combiner.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+func combinePolicies(nps []*networkingv1.NetworkPolicy) *networkingv1.NetworkPolicy {
+	// policies should be combined in alphabetical order
```
Why? It may be worth documenting the reason here.

jackkleeman replied:
Because they are combined in this way by calico-node, when two policies have the same order then they are sorted alphabetically

jackkleeman replied:
I'll comment

robbiev commented on pkg/controllers/networkpolicy/policy_combiner.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+func combinerIndexFunc(obj interface{}) ([]string, error) {
+	if ann, ok := obj.(*networkingv1.NetworkPolicy).Labels[dynamicLabel]; ok {
```
This looks strange, I didn't know that would work to cast it and then get something out of that.

jackkleeman replied:
It does work and it panics if obj is a different type, but we have some guarantees from the informer that this isn't possible

robbiev replied:
As we discussed, I mistakenly thought this `ann, ok` was the result of a safe type cast, but in fact it's the result of the Labels[dynamicLabel] map access.

robbiev commented on pkg/controllers/networkpolicy/policy_combiner.go:
```diff
@@ -0,0 +1,61 @@
+	return nps
+}
+
+func combinerIndexFunc(obj interface{}) ([]string, error) {
```
So the docs for the indexing function (https://godoc.org/k8s.io/client-go/tools/cache#IndexFunc) say
>IndexFunc knows how to compute the set of indexed values for an object.

So does this mean that for this key (the NetworkPolicy object) we are indexing the value of `dynamicLabel`. Which I guess will be the name of the "group" of policies? This doesn't make sense to me.

jackkleeman replied:
My thinking is we have something like `monzo.com/dynamic-networkpolicy: ingress.s-ledger` which will produce a calico network policy called `knp.default.ingress.s-ledger`, which is produced by combining all network policies with that label. Unlabelled policies are not affected.

robbiev replied:
As we discussed, what we're returning here are actually the keys by which we want to index the given obj. It makes sense to me now.

robbiev reviewed (commented):
I had a look through, I'm not super familiar with writing controllers but this looks like a good approach. I left some implementation questions.

jackkleeman committed:
Add rule combiner and associated tests

It seems silly to produce policies with a ton of rules all referencing
different keys of the same label, as these will become separate iptables
rules and ipsets. Instead, we should combine these rules to use
MatchExpressions and the 'in' operator. I have not attempted to cover
complex cases eg with namespace selectors or ip ranges or multiple labels.
